### ISSUE TITLE
Is `legacy_signature_hash` supporting arbitrary sighash types actually useful?

### ISSUE THREAD

**Issue by Kixunil** [rust-bitcoin/rust-bitcoin#3011]
From my understanding, yes, you can append arbitrary `u32` but then you can't serialize it without losing information and if you serialize it with losing information the validating node will verify the signature using a different value thus considering the signature invalid. IOW inputting anything that doesn't fit a `u8` is just producing garbage.

Why don't we already ban this? This makes potential changes to the signing API much harder. (Do I just cast the `u32` to `u8`? What's the point?) Unless I'm missing something we should just ban it entirely since it's also a potential footgun.

**Comment by apoelstra** [comment]
Changing the `u32` value is how you support forkcoins like bcash.

But yes, if we want to support that (supporting it was questionable even when it represented a 10%+ profit vector for Bitcoin holders and probably those days are never coming back..) we shouldn't do it in a primary API.

**Comment by apoelstra** [comment]
I believe the actual historical reasons are:

* It's a `u32` because that's what it "is" in an abstract sense, even though on-chain you can't ever access values outside of the `u8`
* It continues to be weakly typed because of the split between PSBT hashtypes and ECDSA hashtypes, which made changing the API hard.

I definitely think we should change it if we can.

**Comment by apoelstra** [comment]
I like that solution. We should also add a sentence or two to the doccomment of `encode_signing_data_to` explaining the weirdness here.

**Comment by jrakibi** [comment]
> I suggest we keep legacy_encode_signing_data_to accepting u32 and change legacy_signature_hash to take the enum.

is this smth we still want to do? currently both `legacy_encode_signing_data_to` and `legacy_signature_hash` take the enum, which support arbitrary u32 values. So `legacy_signature_hash` currently supports also forkcoins/replay-protection not only non-standard sighash.

**Comment by apoelstra** [comment]
> is this smth we still want to do? 

No, I think we've landed on "make the enum able to express everything".

Closed by #6755 (although we now have #6795 to revisit a different approach).
